hyperbahn/app.js

```diff
@@ -113,6 +120,116 @@ Application.prototype.bootstrap = function bootstrap(cb) {
 
 };
 
+Application.prototype.hookupSignals =
+function hookupSignals() {
+    var self = this;
+
+    process.on('SIGTERM', onSigTerm);
+    process.on('SIGINT', onSigInt);
+
+    function onSigTerm() {
+        self.onSigTerm();
+    }
+
+    function onSigInt() {
+        self.onSigInt();
+    }
+};
+
+Application.prototype.onSigTerm =
+function onSigTerm() {
+    var self = this;
+
+    if (self.tchannel.draining) {
+        self.logger.info('got additional SIGTERM while draining', self.tchannel.extendLogInfo({}));
+    } else {
+        self.startDrain();
+    }
+};
+
+Application.prototype.startDrain =
+function startDrain() {
+    var self = this;
+
+    self.drainStart = self.tchannel.timers.now();
+    self.logger.info('got SIGTERM, draining application', self.tchannel.extendLogInfo({}));
+    self.tchannel.drain('shutting down due to SIGTERM',
+                        isReqDrainExempt, drainedThenClose);
+    self.drainDeadlineTimer = self.tchannel.timers.setTimeout(
+        self.deadlineTimedOut, self.drainDeadlineTimeout);
+
+    function isReqDrainExempt(req) {
+        return self.isReqDrainExempt(req);
+    }
+
+    function drainedThenClose() {
+        self.drainedThenClose();
+    }
+};
+
+Application.prototype.onSigInt =
+function onSigInt() {
+    var self = this;
+
+    if (self.tchannel.draining) {
+        self.finishDrain('got SIGINT, drain aborted');
+    } else if (!self.destroyed) {
+        self.logger.info('got SIGINT, destroying application', self.tchannel.extendLogInfo({}));
+        self.destroy();
+    }
+};
+
+Application.prototype.deadlineTimedOut =
+function deadlineTimedOut() {
+    var self = this;
+
+    self.finishDrain('deadline timeout exceeded, closing now');
+};
+
+Application.prototype.drainedThenClose =
+function drainedThenClose() {
+    var self = this;
+
+    if (!self.destroyed) {
+        self.finishDrain('tchannel drained, destroying application');
+    }
+};
+
+Application.prototype.isReqDrainExempt =
+function isReqDrainExempt(req) {
+    var self = this;
+
+    // we only drain relay requests
+    var chan = self.tchannel.subChannels[req.serviceName];
+    var type = chan && chan.handler && chan.handler.type;
+    if (type === 'tchannel.relay-handler') {
+        return false;
+    } else {
+        return true;
+    }
+};
+
+Application.prototype.finishDrain =
+function finishDrain(mess, info) {
+    var self = this;
+
+    self.drainEnd = self.tchannel.timers.now();
+
+    if (!info) {
+        info = {};
+    }
+    var drainDuration = self.drainEnd - self.drainStart;
+    self.clients.statsd.timing('server.drain-time', drainDuration);
+    info.drainDurationMs = drainDuration;
+    info = self.tchannel.extendLogInfo(info);
+
+    self.logger.info(mess, info);
```

Comment:
This level really should be different based on the message. Successful drain is info; timeout or sigints are warns